File: src/main/java/cn/lifay/ui/form/select/SelectElement.java
package cn.lifay.ui.form.select;

import cn.lifay.ui.form.FormElement;
import javafx.scene.control.ChoiceBox;
import javafx.scene.control.Control;

import java.util.List;

/**
 *@ClassName SelectElement
 *@Description 选择下拉框 T-表单实体 R-ChoiceBox实体 注意：具体显示名称由R的toString()决定
 *@Author 李方宇
 *@Date 2023/1/11 15:02
 **/
public abstract class SelectElement<T,R> extends FormElement<T, R> {

    protected ChoiceBox<R> control = new ChoiceBox<>();
    protected SelectElement(String label, String fieldName) {
        super(label, fieldName);
        init();
    }
    protected SelectElement(String label, String fieldName,List<R> items) {
        super(label, fieldName);
        init();
        control.getItems().addAll(items);
    }

    @Override
    public Control control() {
        return control;
    }

    @Override
    public R getValue() {
        return control.getValue();
    }

    @Override
    public void setValue(Object r) {
        control.setValue((R) r);
    }
    public void loadItems(List<R> items){
        control.getItems().clear();
        control.getItems().addAll(items);
    }
}

File: src/main/java/cn/lifay/ui/form/select/SelectElement.kt
package cn.lifay.ui.form.select

import cn.lifay.ui.form.FormElement
import javafx.scene.Node
import javafx.scene.control.ChoiceBox
import java.awt.SystemColor.text
import kotlin.reflect.KMutableProperty1

/**
 *@ClassName SelectElement
 *@Description 选择下拉框-元素
 *@Author lifay
 *@Date 2023/2/5 13:39
 **/
class SelectElement<T, R : Any>(
    r: Class<R>,
    label: String,
    property: KMutableProperty1<T, R?>,
    required: Boolean = false,
    items: Collection<R?>
) :
    FormElement<T, R>(r, label, property,required = required) {

//    protected var control: ChoiceBox<R> = ChoiceBox<R>()

    init {
        init()
        loadItems(items)
    }

    override fun registerGraphic(): Node {
        return ChoiceBox<R>()
    }

    override fun graphic(): ChoiceBox<*> {
        return graphic as ChoiceBox<*>
    }

    override fun get(): R? {
        return graphic().value as R?
    }

    override fun set(v: R?) {
        graphic().value = v
    }

    override fun clear() {
        graphic().value = defaultValue()
    }

    override fun verify(): Boolean {
        return graphic().value == null
    }

    override fun defaultValue(): R? {
        return null
    }

    fun loadItems(items: Collection<R?>) {
        val choiceBox = graphic as ChoiceBox<R>
        choiceBox.items.clear()
        choiceBox.items.addAll(items)
    }
}